File: Stable.java
/*
 * Authors: Isaiah Mann, Kaylynn Crawford, Zoe Kendall
 * Description: Java Implementation of Stable Marriage problem
 * Adapted from: C++ code by Samia Qader
 */

import java.util.Scanner;
import java.io.FileReader;
import java.io.FileNotFoundException;

class Stable {

  // Markers to denote when men and women appear in the text file
  static final String menMarkerInFile = "M";
  static final String womenMarkerInFile = "W";

  public static void main (String[] args) {
      Scanner in = null;

      // Create Input variables to hold numberOfEachGender ber of choicesOfMen and choicesOfWomen:
      int choices =0, numberOfEachGender =0, n =1, i, j;
      int [][] choicesOfMen = null, choicesOfWomen = null;

      // Print a nice intro the user user
      System.out.println("~~~~Welcome to the Marriage Algorithm Program~~~~");
      System.out.println("The rules for this program are as follows:");
      System.out.println("1. Enter the choices for each man/woman");
      System.out.println("   The choices must be enterred with a space between each value");
      System.out.println("2. Do not enter a choice that is not on your list");
      System.out.println("   This program does not check for such errors");
      System.out.println("3. Do not repeat a choice for the same man/woman");
      System.out.println("4. Complete each Rank ordered List (ROL)\n\n");

       if(args.length == 0) {
         in = new Scanner(System.in);

          // Not sure what this five thing is about, probably worth looking into
          // Read in the numberOfEachGender ber of choicesOfMen/choicesOfWomen
          System.out.println("How many men/women ?");

          numberOfEachGender = in.nextInt();

          // Ensure the numberOfEachGender ber is no greater than five
          while (numberOfEachGender  > 5) {
             System.out.println("Please don't enter a number greater than 5");
             System.out.println("Re-enter number:");
             numberOfEachGender  = in.nextInt();
          }


          // Create two jagged arrays, one for choicesOfMen, one for choicesOfWomen
          choicesOfMen = new int[numberOfEachGender ][numberOfEachGender +1];
          choicesOfWomen = new int[numberOfEachGender ][numberOfEachGender +1];


          //Read choices for choicesOfMen
          System.out.println("Enter your choices for: ");
          for (i = 0; i < numberOfEachGender ; i++) {

              System.out.println("Man " + i + ": ");


              for (j = 0; j < numberOfEachGender ; j++) {
                  choices = in.nextInt();

                  while (choices >= numberOfEachGender  || choices < 0) {
                    System.out.println("Please enter a valid choice for Man " + i + " between 0 and " + (numberOfEachGender -1) + ": ");
                    choices = in.nextInt();
                  }

                  choicesOfMen[i][j+1] = choices;
              }

            }

            System.out.println("\n");



            // Read choices for choicesOfWomen
            for (i = 0; i < numberOfEachGender ; i++) {
                System.out.println("Woman " + i + ": ");

                for (j = 0; j < numberOfEachGender ; j++) {
                    choices = in.nextInt();

                    while (choices >= numberOfEachGender  || choices < 0) {
                      System.out.println("Please enter a valid choice for Woman " + i + " between 0 and " + (numberOfEachGender -1) + ": ");
                      choices = in.nextInt();
                    }
                    choicesOfWomen[i][j+1] = choices;
                }
            }
          } else {

              try {
                in = new Scanner(new FileReader(args[0]));
              } catch (FileNotFoundException e) {
                System.out.println("File: " + args[0] + " not found. Exiting program...");
                System.exit(0);
              }

              numberOfEachGender = Integer.parseInt(in.nextLine());

              choicesOfMen = new int [numberOfEachGender][numberOfEachGender + 1];
              choicesOfWomen = new int [numberOfEachGender][numberOfEachGender + 1];

              int[][] currentChoices = null;

              String marker = in.nextLine();
              // Allows for arbitrary ordering: choicesOfMen or choicesOfWomen can be first based off markers
              if (isWomenMarker(marker)) {
                currentChoices = choicesOfWomen;
              } else if (isMenMarker(marker)) {
                currentChoices = choicesOfMen;
              } else {
                System.out.println("Invalid marker: " + marker + ". Exiting the program...");
                System.exit(0);
              }

              String proceedingLine = readInChoicesFromScanner(currentChoices, in);

              if (isWomenMarker(proceedingLine)) {
                currentChoices = choicesOfWomen;
              } else if (isMenMarker(proceedingLine)) {
                currentChoices = choicesOfMen;
              } else {
                System.out.println("Invalid marker: " + proceedingLine + ". Exiting the program...");
                System.exit(0);
              }

              readInChoicesFromScanner(currentChoices, in);
          }

      int[][] proposers = null;
      int[][] acceptors = null;

      System.out.println("Who should propose:");
      System.out.println("1). The Women?");
      System.out.println("2). The Men?");
      in = new Scanner(System.in);

      int proposerChoice = in.nextInt();

      while (proposerChoice != 1 && proposerChoice != 2) {
        System.out.println("Please enter a valid number (1 or 2)");
        proposerChoice = in.nextInt();
      }

      if (proposerChoice == 1) {
        proposers = choicesOfWomen;
        acceptors = choicesOfMen;
      } else if (proposerChoice == 2) {
        proposers = choicesOfMen;
        acceptors = choicesOfWomen;
      }


      // Actually run the stable matching algorithm:
      //Marriage Algorithm
      int proposer= -1, acceptor= -1, uglyProposer = numberOfEachGender  + 1, temp;
      int indexOfProposer = -1, indexOfAcceptor = -1;
      int [] indexesOfAcceptors = new int [numberOfEachGender];
      int couples = 0, indexer = 1;

      for (i = 0; i < numberOfEachGender ; i++) {
          indexesOfAcceptors[i] = uglyProposer;
      }

      while(couples < numberOfEachGender ) {

         proposer = couples;
         while (proposer != uglyProposer) {
           // used later
           // acceptor = best choice on proposer's list
           while (acceptor == -1) {
              if (proposers[proposer][indexer]!= -1) {
                  indexOfAcceptor = indexer;
                  acceptor = proposers[proposer][indexer];
              } else {
                  indexer++;
              }
            }// end while acceptor

            // index of acceptor on proposer's list
            for (i = 0; i <= numberOfEachGender ; i++) {
                if (acceptors[acceptor][i]== proposer) {
                  indexOfProposer = i;
                }
            }

            // if acceptor prefers proposer to her fiance
            if (indexOfProposer < indexesOfAcceptors[acceptor]) {
                temp = indexesOfAcceptors[acceptor]; // temp = old fiance
                indexesOfAcceptors[acceptor] = indexOfProposer; // fiance = new proposer
                indexOfProposer = temp; // proposer = old fiance

                if (indexOfProposer == uglyProposer) {
                    proposer = uglyProposer;  // proposer = ugly proposer
                } else {
                    proposer = acceptors[acceptor][indexOfProposer];  // proposer = old fiance
                }

            } // end if indexOfProposer

            if (proposer != uglyProposer) {
              proposers[proposer][indexOfAcceptor] = -1;
            }

            acceptor = -1;
            indexer = 1;
         }// end while indexOfProposer

       couples++;

      }// end while couples
       // withdraw w from m's list
      //set the acceptor = -1 (no one)
      //increment the numberOfEachGender ber of couples
      System.out.println("\n\n");

      // Print out the reuslts:
      System.out.println("~~~~~~~~RESULTS~~~~~~~~");
      if (proposerChoice == 1) {
        for (i = 0; i < couples; i++) {
          System.out.println("Couple " + (i+1) + ": Woman " + choicesOfMen[i][indexesOfAcceptors[i]] + " & Man " + i + "\n");
        }
      } else if (proposerChoice == 2) {
        for (i = 0; i < couples; i++) {
          System.out.println("Couple " + (i+1) + ": Man " + choicesOfWomen[i][indexesOfAcceptors[i]] + " & Woman " + i + "\n");
        }
      }

      //end main
  }

  // Returns the last line it reads in (presumably a marker for choicesOfMen or choicesOfWomen)
  static String readInChoicesFromScanner (int[][] choices, Scanner scanner) {
      String nextLine = "";
      int personIndex = 0;
      while (scanner.hasNextLine() && !isMarker(nextLine = scanner.nextLine())) {

        choices[personIndex] = readInChoicesFromLine(nextLine);

        personIndex++;

      }

      return nextLine;
  }

  static int[] readInChoicesFromLine (String commaSeparatedLine) {
    String[] choicesAsString = commaSeparatedLine.split(",");

    // Allow for an offset of one to match the current style
    int[] choices = new int [choicesAsString.length + 1];

    for (int i = 0; i < choicesAsString.length; i++) {
      choices[i + 1] = Integer.parseInt(choicesAsString[i]);
    }

    return choices;

  }

  static boolean isWomenMarker (String marker) {
    return (marker.equals(womenMarkerInFile));
  }

  static boolean isMenMarker (String marker) {
    return (marker.equals(menMarkerInFile));
  }

  static boolean isMarker (String marker) {
    return isMenMarker(marker) || isWomenMarker(marker);
  }
}

File: stable.py
# Authors: Isaiah Mann, Kaylynn Crawford, Zoe Kendall
# Description: Python Implementation of Stable Marriage problem
# Adapted from: C++ code by Samia Qader

import re
import sys

# UTILITY METHODS:

# Code adapted from http://stackoverflow.com/questions/2397141/how-to-initialize-a-two-dimensional-array-in-python
def create_2d_int_array (x, y):
    return [[0 for i in range(y)] for j in range(x)]

def next_int (message = None):
    if (message == None):
        return int(input())
    else:
        return int(input(message))

def read_int_from_file (inputFile):
    return int(inputFile.readLine())

def is_women_marker (marker):
    return marker == women_marker_in_file

def is_men_marker (marker):
    return marker == men_marker_in_file

def is_marker (marker):
    return (is_women_marker(marker) or
            is_men_marker(marker))

def read_in_choices_from_line (comma_separated_line):
    choices_as_string = comma_separated_line.split(",")

    choices = [i for i in range(len(choices_as_string) + 1)]

    for i in range(0, len(choices_as_string)):
        choices[i + 1] = int(choices_as_string[i])

    return choices;

def read_in_choices_from_file (choices, inputFile):
    next_line = ""
    person_index = 0

    next_line = inputFile.pop(0)
    while (bool(next_line) and not is_marker(next_line)):
        choices[person_index] = read_in_choices_from_line(next_line)

        person_index += 1
        next_line = inputFile.pop(0)

    return next_line


# Reading in Input:
choices = 0
numberOfEachGender = 0
n = 0
i = 0
j = 0

# Markers to denote when men and women appear in the text file
men_marker_in_file = "M";
women_marker_in_file = "W";

print("~~~~Welcome to the Marriage Algorithm Program~~~~")
print("The rules for this program are as follows:")
print("1. Enter the choices for each man/woman")
print("   The choices must be enterred with a space between each value")
print("2. Do not enter a choice that is not on your list")
print("   This program does not check for such errors")
print("3. Do not repeat a choice for the same man/woman")
print("4. Complete each Rank ordered List (ROL)\n\n")

# If there is no file passed to the program
if(len(sys.argv) <= 1):

    numberOfEachGender = next_int("How many men/women ?\n")
    while (numberOfEachGender > 5):
        print("Please don't enter a number greater than 5")
        numberOfEachGender = next_int("Re-enter number:\n")

    choices_of_men = create_2d_int_array(numberOfEachGender, numberOfEachGender+1)
    choices_of_women = create_2d_int_array(numberOfEachGender, numberOfEachGender+1)

    for i in range (0, numberOfEachGender):
        print ("Man " + str(i) + ": ")

        for j in range (0, numberOfEachGender):
            choices = next_int()

            while (choices >= numberOfEachGender or choices < 0):
                print("Please enter a valid choice for Man " + str(i) + " between 0 and " + str(numberOfEachGender-1) + ": ")
                choices = next_int()
            choices_of_men[i][j+1] = choices

    for i in range (0, numberOfEachGender):
        print ("Woman " + str(i) + ": ")

        for j in range (0, numberOfEachGender):
            choices = next_int()

            while (choices >= numberOfEachGender or choices < 0):
                print("Please enter a valid choice for Woman " + str(i) + " between 0 and " + str(numberOfEachGender-1) + ": ")
                choices = next_int()
            choices_of_women[i][j+1] = choices
else:
    # Opens the input file in read mode
    input_file = open(sys.argv[1], 'r')

    input_text = input_file.read()

    input_lines = input_text.split("\n")

    numberOfEachGender = int(input_lines.pop(0))

    choices_of_women = create_2d_int_array(
        numberOfEachGender,
        numberOfEachGender + 1
    )

    choices_of_men = create_2d_int_array(
        numberOfEachGender,
        numberOfEachGender + 1
    )

    current_choices = None

    marker = input_lines.pop(0)

    if (is_women_marker(marker)):
        current_choices = choices_of_women
    elif (is_men_marker(marker)):
        current_choices = choices_of_men
    else:
        print("Invalid Marker: " + marker + ". Exiting program...")
        sys.exit(0)

    marker = read_in_choices_from_file(current_choices, input_lines)

    if (is_women_marker(marker)):
        current_choices = choices_of_women
    elif (is_men_marker(marker)):
        current_choices = choices_of_men
    else:
        print("Invalid Marker: " + marker + ". Exiting program...")
        sys.exit(0)

    read_in_choices_from_file(current_choices, input_lines)

proposers = None
acceptors = None

print ("Who should propose:")
print ("1). The Women?")
print("2). The Men?")
proposer_choice = next_int()

while (proposer_choice != 1 and proposer_choice != 2):
    print("Please enter a valid number (1 or 2)")
    proposer_choice = next_int()

if (proposer_choice == 1):
    proposers = choices_of_women
    acceptors = choices_of_men
elif(proposer_choice == 2):
    proposers = choices_of_men
    acceptors = choices_of_women

#  Run Stable Marriage Algorithm:
current_proposer = -1
current_acceptor = -1
ugly_proposer = numberOfEachGender + 1
temp = 0
index_of_proposer = -1
index_of_acceptor = -1
#Initializer adapated from: http://stackoverflow.com/questions/521674/initializing-a-list-to-a-known-numberOfEachGenderber-of-elements-in-python
indexes_of_acceptors = [0] * numberOfEachGender
couples = 0
indexer = 1


for i in range (0, numberOfEachGender):
    indexes_of_acceptors[i] = ugly_proposer

while (couples < numberOfEachGender):

    current_proposer = couples
    while (current_proposer != ugly_proposer):

        while (current_acceptor == -1):
            if (proposers[current_proposer][indexer] != -1):
                index_of_acceptor = indexer
                current_acceptor = proposers[current_proposer][indexer]
            else:
                indexer += 1

        for i in range(0, numberOfEachGender+1):
            if (acceptors[current_acceptor][i] == current_proposer):
                index_of_proposer = i

        if (index_of_proposer < indexes_of_acceptors[current_acceptor]):
            temp = indexes_of_acceptors[current_acceptor]
            indexes_of_acceptors[current_acceptor] = index_of_proposer
            index_of_proposer = temp

            if (index_of_proposer == ugly_proposer):
                current_proposer = ugly_proposer
            else:
                current_proposer = acceptors[current_acceptor][index_of_proposer]

        if (current_proposer != ugly_proposer):
            proposers[current_proposer][index_of_acceptor] = -1

        current_acceptor = -1
        indexer = 1

    couples += 1
print("\n\n")
#       // Print out the reuslts:
print("~~~~~~~~RESULTS~~~~~~~~")
if (proposer_choice == 1):
    for count in range(0, couples):
        print("Couple " + str(count+1) + ": Woman " + str(choices_of_men[count][indexes_of_acceptors[count]]) + " & Man " + str(count) + "\n")
elif(proposer_choice == 2):
    for count in range(0, couples):
        print("Couple " + str(count+1) + ": Man " + str(choices_of_women[count][indexes_of_acceptors[count]]) + " & Woman " + str(count) + "\n")
